[[ Bug 23085 ]] Fix Macos Big Sur visual effect rendering

livecodeian opened the pull request:
This patch revises the fix applied for bug 23019 to resolve a problem introduced by that patch. This changes the implementation of MCMacPlatformWindow::DoUpdate to remove the call to NSView method displayIfNeededInRect and allow window redraws to occur while in the NSApp event loop.

livecodeian committed:
[[ Bug 23085 ]] Fix Macos Big Sur visual effect rendering

This patch revises the fix applied for bug 23019 to resolve a problem
introduced by that patch. This changes the implementation of
MCMacPlatformWindow::DoUpdate to remove the call to NSView method
displayIfNeededInRect and allow window redraws to occur while in the
NSApp event loop.

livecodeian committed:
[[ Bug 23085 ]] Add release note for bug 23085

This patch adds a release note for bug 23085, where visual effects
would not render correctly on Macos Big Sur

runrevmark commented on engine/src/mac-window.mm:
```diff
@@ -1588,6 +1588,8 @@ - (void)drawRect: (NSRect)dirtyRect
 	
 	MCGRegionDestroy(t_update_region);
+	
+	MCMacPlatformSyncUpdateAfterDraw(self.window.windowNumber);
```
I take it this forces the event sifting in DoUpdate to break? (Perhaps a comment to that effect would be useful :) ).

livecodeian replied:
It does, I've added a comment.

runrevmark commented on engine/src/mac-window.mm:
```diff
@@ -2216,7 +2224,15 @@ bool MCMacDoUpdateRegionCallback(void *p_context, const MCRectangle &p_rect)
+	{
+		NSEvent *t_event;
+		t_event = [NSApp nextEventMatchingMask: NSApplicationDefinedMask
+									 untilDate: [NSDate dateWithTimeIntervalSinceNow: 0.1]
```
I take it we can't get away with a smaller interval (maybe even zero?) - and if we can, do we still need the sync event?

livecodeian replied:
I've tested with lowering the interval - 0.02 ms seems to fix the redraw issues. Lower than that brings them back. That would be consistent with a maximum redraw rate of 60hz

runrevmark replied:
@livecodeian The interval here is still 0.1s - but the above statement suggests 0.02 works... Additionally I suggest explaining why 0.02s in the comment above :)

runrevmark reviewed (commented):
Can you see if you can measure the effect on performance of this patch (on non-Big Sur)? i.e. how long it takes to update a field N times (no lock screen) with the old method vs the new method.

livecodepanos commented:
Also this needs to be rebased against `release-9.6.2`

livecodeian commented:
I've measured the update speed - around 0.043ms with this fix vs 0.987ms without so this method seems to be actually faster than previous.

runrevmark commented:
> I've measured the update speed - around 0.043ms with this fix vs 0.987ms without so this method seems to be actually faster than previous.

Wow - that's quite an improvement... Is that just on Big Sur, or Catalina too?

livecodeian commented:
@runrevmark branch rebased and retargetted against release-9.6.2

livecodeian committed:
[[ Bug 23085 ]] Wait until redraw before continuing from update

This patch modifies MCMacPlatformWindow::DoUpdate to wait for drawRect
to be called on the underlying system view. This ensures that the
redraw has been performed as requested.

livecodeian committed:
[[ Bug 23085 ]] Check if sheet is showing when waiting for window update

This patch modifies MCMacPlatformWindow::DoUpdate to avoid waiting in a
loop for the call to drawRect if the update occurs while showing a modal
dialog as a sheet.

livecodeian committed:
[[ Bug 23085 ]] Limit redraw fix to Macos BigSur

This patch modifies the previous changes to wait for redraws within
MCMacPlatformWindow::DoUpdate so that they only apply when running on
Macos Big Sur.

runrevmark commented on engine/src/mac-window.mm:
```diff
@@ -2214,10 +2238,32 @@ bool MCMacDoUpdateRegionCallback(void *p_context, const MCRectangle &p_rect)
+		// Frequent redraws with displayIfNeeded causes graphical glitches on Macos Big Sur, so instead
+		// we enter the runloop to trigger a redraw. This will cause drawRect to be invoked on our view
+		// which in turn will result in a redraw window callback being sent.
+		// The timeout value of 0.02ms is specified to avoid hitting the 60hz redraw limit.
```
This comment is now out of date relative to the constant 2.0 used below (which I believe is seconds?)

livecodeian replied:
The edit from 0.02 to 2.0 was a mistake which I have now corrected :/

runrevmark commented on engine/src/mac-internal.h:
```diff
@@ -542,6 +544,8 @@ class MCMacPlatformWindow: public MCPlatformWindow
     // The window's content view.
     MCWindowView *m_view;
+	
+	bool m_waiting_for_draw;
```
I think this would be fine put in the bitset struct below?

livecodeian replied:
Yes, I've moved it now

runrevmark commented:
@livecode-vulcan review ok c5a4759

livecode-vulcan commented:
💙 review by @runrevmark ok c5a4759ba6a584c354b4804685abae0d312353f0

livecode-vulcan commented:
😎 test success c5a4759ba6a584c354b4804685abae0d312353f0

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/974)
* try-community-armv7-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-ndk16r15/builds/382)
* try-community-arm64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-ndk16r15/builds/381)
* try-community-x86-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-ndk16r15/builds/380)
* try-community-x86_64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-ndk16r15/builds/381)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2499)
* try-community-universal-ios-iphoneos14.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos14.4/builds/28)
* try-community-universal-ios-iphoneos14.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos14.1/builds/77)
* try-community-universal-ios-iphonesimulator14.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator14.4/builds/28)
* try-community-universal-ios-iphonesimulator14.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator14.1/builds/75)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2526)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2515)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2513)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2550)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2486)
